feat!: make DhtHeader field non-malleable

philipr-za opened the pull request:
Description
---
This PR includes the relevant DhtHeader fields in the commitment of a Dht Message MAC signature so that they cannot be changed while in route.

The header fields included in the commitment are:
- Major version
- Minor version
- Destination
- Message type
- Message flags
- Expiry time (if exists)
- Ephemeral public key (if exists)

How Has This Been Tested?
---
Rust integration test provided and existing tests updated.

BREAKING CHANGE: Previous message signatures will not be valid

delta1 commented on comms/dht/src/inbound/decryption.rs:
```diff
@@ -193,8 +193,17 @@ where S: Service<DecryptedDhtMessage, Response = (), Error = PipelineError>
     ) -> Result<DecryptedDhtMessage, DecryptionError> {
         let dht_header = &message.dht_header;
 
+        let mut header_mac_bytes = dht_header.major.to_le_bytes().to_vec();
```
minor perf: is it worth allocating this vec with capacity before? Maybe not necessary without a profile but there would be a few reallocations.

philipr-za replied:
I guess you might as well

delta1 commented on comms/dht/src/test_utils/makers.rs:
```diff
@@ -75,14 +75,28 @@ pub fn make_dht_header(
     flags: DhtMessageFlags,
     include_origin: bool,
     trace: MessageTag,
+    include_destination: bool,
```
lil clippy warning here

philipr-za replied:
damnit I forget to check the tests locally

delta1 commented on comms/dht/tests/dht.rs:
```diff
@@ -629,6 +630,99 @@ async fn dht_propagate_message_contents_not_malleable_ban() {
+    }
+
+    let out_msg = Person {
+        name: "John Conway".into(),
```
RIP 🙏

philipr-za replied:
damnit, I forget to check the tests locally

delta1 reviewed (dismissed):
Looking great, happy to approve with clippy passing

delta1 reviewed (dismissed):
LGTM

stringhandler reviewed (commented):
Happy to merge this once a WIREMODE update is included

philipr-za committed:
feat!: Make DhtHeader field non-malleable

### Breaking Change to comms

This PR includes the relevant DhtHeader fields in the commitment of a Dht Message MAC signature so that they cannot be changed while in route.

The header fields included in the commitment are:
- Major version
- Minor version
- Destination
- Message type
- Message flags
- Expiry time (if exists)
- Ephemeral public key (if exists)

philipr-za commented:
> Happy to merge this once a WIREMODE update is included

Ahh yes you are correct I should have incremented the DHT_MAJOR_VERSION and have updated that.

sdbondi reviewed (approved):
utAck - obviously would be nice to DRY up the challenge construction
